File: dfs/105.construct-binary-tree-from-preorder-and-inorder-traversal.go
package leetcode

// @leet start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func dfsbuild(preStart, inStart, length int, preorder []int, inorderM map[int]int) *TreeNode {
	if length == 0 {
		return nil
	}
	rootVal := preorder[preStart]
	inorderNeedle := inorderM[rootVal]
	leftSize := inorderNeedle - inStart
	rightSize := length - leftSize - 1

	l := dfsbuild(preStart+1, inStart, leftSize, preorder, inorderM)
	r := dfsbuild(preStart+1+leftSize, inorderNeedle+1, rightSize, preorder, inorderM)
	newNode := &TreeNode{
		Val:   rootVal,
		Left:  l,
		Right: r,
	}
	return newNode
}

func buildTree(preorder []int, inorder []int) *TreeNode {
	inorderM := make(map[int]int)
	for i, v := range inorder {
		inorderM[v] = i
	}
	newTree := dfsbuild(0, 0, len(inorder), preorder, inorderM)
	return newTree
}

// @leet end

File: dfs/105.construct-binary-tree-from-preorder-and-inorder-traversal.ts
import { TreeNode } from "./type";

// @leet start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function buildTree(preorder: number[], inorder: number[]): TreeNode | null {
  const inorderM = new Map<number, number>();
  for (const [idx, val] of inorder.entries()) {
    inorderM.set(val, idx);
  }

  function dfsBuild(
    preStart: number,
    inStart: number,
    length: number,
  ): TreeNode | null {
    if (length === 0) return null;
    const rootVal = preorder[preStart];
    const inorderNeedle = inorderM.get(rootVal)!;
    const leftSize = inorderNeedle - inStart;
    const rightSize = length - leftSize - 1;

    return new TreeNode(
      rootVal,
      dfsBuild(preStart + 1, inStart, leftSize),
      dfsBuild(preStart + 1 + leftSize, inorderNeedle + 1, rightSize),
    );
  }
  return dfsBuild(0, 0, preorder.length);
}
// @leet end

